File: day7/findinRotatedArray.java
//binary search
import java.util.*;
public class findinRotatedArray {
    public int search(int[] nums, int target) {
        int l=0;int h=nums.length-1;
        while(l<=h){
            int mid=l+(h-l)/2;
            if(nums[mid]==target)return mid;

            else if(nums[l]<=nums[mid]){
                if(nums[l]<=target && nums[mid]>target){
                    h=mid-1; }
                else{
                 l=mid+1;
                }
            }
            else{
                if(nums[mid]<target && target<=nums[h] ){
                    l=mid+1;
                }
                else{
                    h=mid-1;
                }
            }
        }
        return -1;
    }
   public void main(String[] st){
    Scanner sc = new Scanner(System.in);
    int n=sc.nextInt();
    int arr[]=new int[n];
    for(int i=0;i<n;i++){
        arr[i]=sc.nextInt();
    }
    int target = sc.nextInt();
    System.out.print(search(arr, target));
   }
}

File: binary_search/findinRotatedArray.cpp
//#Binary Search
//Q- a sorted rotated array is given we have to find a variable
#include<iostream>
using namespace std;
#include<vector>

int findArrayInRotatedArray(vector<int> arr,int target){
    int s=0,e= arr.size()-1;
    while(s<=e){
        int mid=s+(e-s)/2;
        if(arr[mid]==target) return mid;
        else if(arr[mid]>arr[e]){
                if(arr[s]<= target && arr[mid]> target) e=mid-1;
                else s=mid+1;
        } else{
            if(target>arr[mid] && target<=arr[e] ) s=mid+1;
            else e=mid-1;
        }
    }
    return -1;
}

int main(){
    int n;
    cin>>n;
    vector<int> arr(n);
    for(int i=0;i<n;i++) {cin>>arr[i];}
    int target;
    cin>>target;
    cout<< findArrayInRotatedArray(arr,target);
}